metrics/sinks/hawkular/driver_historical.go

```diff
@@ -0,0 +1,432 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hawkular
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/hawkular/hawkular-client-go/metrics"
+	"k8s.io/heapster/metrics/core"
+)
+
+// Historical returns the historical data access interface for this sink
+func (h *hawkularSink) Historical() core.HistoricalSource {
+	return h
+}
+
+// GetMetric retrieves the given metric for one or more objects (specified by metricKeys) of
+// the same type, within the given time interval
+func (h *hawkularSink) GetMetric(metricName string, metricKeys []core.HistoricalKey, start, end time.Time) (map[core.HistoricalKey][]core.TimestampedMetricValue, error) {
+	typ := h.metricNameToHawkularType(metricName)
+
+	resLock := &sync.Mutex{}
+	res := make(map[core.HistoricalKey][]core.TimestampedMetricValue, len(metricKeys))
+
+	errChan := make(chan error, len(metricKeys))
+	wg := &sync.WaitGroup{}
+
+	for _, keyM := range metricKeys {
+		wg.Add(1)
+		go func(key core.HistoricalKey) {
+			defer wg.Done()
+			tags := h.keyToTags(&key)
+			tags[descriptorTag] = metricName
+
+			o := []metrics.Modifier{metrics.Filters(metrics.TagsFilter(tags), metrics.TypeFilter(typ))}
+			if h.isNamespaceTenant() && key.NamespaceName != "" && key.ObjectType != core.MetricSetTypeCluster {
+				o = append(o, metrics.Tenant(key.NamespaceName))
+			}
+
+			// Remove in favor of using ReadRaw with filters to reduce the amount of queries to HWKMETRICS,
+			// but that API is still considered unstable in current release
+			mds, err := h.client.Definitions(o...)
+			if err != nil {
+				errChan <- err
+				return
+			}
+
+			if len(mds) > 1 {
+				errChan <- fmt.Errorf("Given metric query (metricName=%q, key=%q) did not result in unique metricId, found %d results", metricName, key, len(mds))
+				return
+			} else if len(mds) < 1 {
+				return
+			}
+
+			filters := make([]metrics.Filter, 0, 2)
+
+			// Without start & endTime, the default is requesting 8 hours to the past
```

Comment:
Agreed. The 8 hours is the default in Hawkular when user gives no range restrictions, but I'd much rather have some application specific restriction here. Should it be added to the API level that if both are empty, the API would set them with some default? Instead of doing the same filtering on each historicalSource?